lib/types.nix

```diff
@@ -466,7 +466,11 @@ rec {
       descriptionClass = "noun";
       check = x: isDerivation x || isStorePath x;
       merge = loc: defs:
-        let res = mergeOneOption loc defs;
+        # See https://github.com/NixOS/nixpkgs/issues/222753 for design details.
+        # Reconcile packages using `outPath` equality
+        # `meta` attribute is ignored as it is not supposed
+        # to have any role in the actual contents of the derivation.
+        let res = mergeEqualOption loc (map (def: def.outPath) defs);
```

Comment:
is `toDerivation` no-op on existing derivations or stuff that possess `outPath` or should I actively filter and map only those who need to be transformed?

Reply:
This does not work for me, `def` is not the actual derivation, it's an attrset with attributes `file` and `value`, so we'd need to look at `def.value.outPath`.

The bigger issue though, is that mapping `def.value` to the `outPath` changes the value that we get from the option in the module system, we would get a string (the `outPath`), instead of the actual derivation that was assigned.

I think we could solve this by introducing something like `mergeEqualOptionBy` that takes a comparator function to compare the values with and decide if they are equal, something like

```nix
  mergeEqualOptionBy = areEqual: loc: defs:
    if defs == [] then abort "This case should never happen."
    # Return early if we only have one element
    # This also makes it work for functions, because the foldl' below would try
    # to compare the first element with itself, which is false for functions
    else if length defs == 1 then (head defs).value
    else (foldl' (first: def:
      if (areEqual def.value first.value) then
        throw "The option `${showOption loc}' has conflicting definition values:${showDefs [ first def ]}\n${prioritySuggestion}"
      else
        first) (head defs) (tail defs)).value;

```

and then we can redefine `mergeEqualOption` as `mergeEqualOptionBy lib.id`.

Reply:
Feel free to adopt this PR and bring it to completion, I am out of time on this stuff alas…